[modbus.lambda] Lambda Heat Pump

chilobo opened the pull request:
This is my first contribution on github!

Based on the Stiebel Eltron bundle by Paul Frank I contribute a bundle to manage Lambda Heat Pumps as a OSG addon to the Modbus Binding.

I did some testing by adding my jar to the local addon directory. 
See the README.md for restrictions on usage.

The binding is based on the Modbus description of the manufacturer:
(https://lambda-wp.at/wp-content/uploads/2022/01/Modbus-Protokoll-und-Beschreibung.pdf)

Compressed directory:
[org.openhab.binding.modbus.lambda.zip](https://github.com/user-attachments/files/17457153/org.openhab.binding.modbus.lambda.zip)

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unusable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

All PRs should be created using the "main" branch as base.
Important bugfixes are cherry-picked by maintainers to the patch release branch after a PR has been merged.

Add one or more appropriate labels to make your PR show up in the release notes.
E.g. enhancement, bug, documentation, new binding
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst

# Title

Provide a short summary in the *Title* above. It will show up in the release notes.
For example:
- [homematic] Improve communication with weak signal devices
- [timemachine][WIP] Initial contribution

# Description

Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work

# Testing

Your pull request will automatically be built and available under the following folder:
https://openhab.jfrog.io/ui/native/libs-pullrequest-local/org/openhab/addons/bundles/

It is a good practice to add a URL to your built JAR in this pull request description,
so it is easier for the community to test your Add-on.
If your pull request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Marketplace:
https://community.openhab.org/c/marketplace/69

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

lsiepel commented:
Thank you for this contribution. It would be nice if you could perform a self review see the checklist here: https://github.com/openhab/openhab-addons/wiki/Review-Checklist
Also make sure that this binding is also added to files outside of the binding folder:
bom/openhab-addons/pom.xml
bundles/pom.xml
CODEOWNERS

Once these are done, me or another maintainer will review this. Just ping if you need anything.

chilobo commented:
> Once these are done, me or another maintainer will review this. Just ping if you need anything.
I added the changes to 
bom/openhab-addons/pom.xml
bundles/pom.xml
and put some // in front of my trace loggings.
I read the Review-Checklist. 
I guess some of the channel names are too long, but
- I tried to stick with the names the manufacturer provided
- Some descriptions of the modbus registers are quite similar in different groups, so shortening them could produce legibility

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/java/org/openhab/binding/modbus/lambda/internal/handler/LambdaException.java:
```diff
@@ -0,0 +1,33 @@
+ * @author Paul Frank - Initial contribution
+ * @author Christian Koch - modified for lambda heat pump based on stiebeleltron binding for modbus
+ */
+@SuppressWarnings("serial")
```
Please don't suppress, but handle the case.

chilobo replied:
I do not have a Heat Pump connected by serial modbus, I think this mode of connection is not used because the heat pumps of lambda are connected by TCP anyway. So I would rather leave it like it is.
Should I add a comment in the README.md?

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/config/config-descriptions.xml:
```diff
@@ -0,0 +1,25 @@
+
+	<config-description uri="thing-type:lambda:modbusconfig">
+
+		<parameter name="refresh" type="integer" min="1" unit="s">
```
Please make name, label, description (and readme.md) consistent.
1. is it Poll Interval or Refresh Interval ? 
2. Is the min 1 (xml attribute) or zero (description) ?
3. Is the right default is 20 (here) or 30 (readme) ?

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/config/config-descriptions.xml:
```diff
@@ -0,0 +1,25 @@
+		</parameter>
+
+		<parameter name="maxTries" type="integer" min="1">
+			<label>Maximum Tries When Reading</label>
```
```suggestion
			<label>Maximum Tries</label>
```

lsiepel reviewed (changes requested):
Thank you for contributing to openHAB, greatly appreciated. This is a partly review, I looked at 33/38 files. I left some comments, the most important one is about getting the openHAB side of the moddeling right. From there on i can look at the rest of the files. Let me know if you need anything

chilobo commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+
+
+
+## Full Example
```
I changed the tabs in my editor to spaces. 
I already supplied a sketchy example how to use the E-Manager function of the heat pump.

lsiepel replied:
As per template:
_Provide a full usage example based on **textual configuration files**._

Currently i see a yaml example from the UI.

chilobo replied:
I started to split the bundle in seperate ones so that a user can use more than one boiler, buffer, heating circuit and heatpump. But I ran into problems with the handlers. I am not sure if that is a naming problem. 
Anyway I will not have time to correct your above mentioned issues within the next **two weeks**.

Anyway: Can you suggest one or two more modern bundles, preferably modbus, where I can learn how to improve my code, especially:
- 
When I split my bundle into 5 different ones, do I have to organize them in 5 different directories
openhab-addons\bundles\org.openhab.binding.modbus.lambdageneral\
openhab-addons\bundles\org.openhab.binding.modbus.lambdabuffer\
...
openhab-addons\bundles\org.openhab.binding.modbus.lambdaheatingcircuit\
or something like
openhab-addons\bundles\org.openhab.binding.modbus.lambdaheating\general\
...
openhab-addons\bundles\org.openhab.binding.modbus.lambdaheating\heatingcircuit\
with common code in 
openhab-addons\bundles\org.openhab.binding.modbus.lambdaheating\   ?

- handling of units;
- handlers;
- seperating parts of the Lambdahandler which is an instance (?) of BaseThingHandler into common usable code by the 5 different handlers I have to program.

lsiepel replied:
Ah, now understand your hesitation. We need to go back one step. There is no need to create seperate bundles. Just different Handlers within the bundle. No need to start fiddling with directory's, the change is much smaller.

1. Just copy paste the current `LambdaHandler.java` to `HeatpumpHandler.java`, `HeatingCircuitHandler.java` etc. Each Thing has its own handler. Obviously the contents of the files have to be stripped down to just the logic from that Thing it handles.
2. Adjust the `LambdaHandlerFactory.java` to return the right `Handler` for the right `Thing`
3. Re-arrange the thing-structure.xml so the rigth channels belong to the right Thing
4. Adjust README.md
5. Probably some improvements can be made with a `BaseThingHandler` to de-duplicate code in the handlers, but as being said that is implementation specific.

The astro binding uses this pattern (moon and sun are differnt things, they share a base handler) There are many more bindings using this pattenr, just open some and have q quick look at the handlers. The only thing that makes the modbus specific is the relation with the base modbus binding for the bridge, but that is not affecting this suggested pattern.
Hopefully this gives some guidance, i would suggest to start with just extract just one group e.g. the boiler (or pick one) and make that fully functional, if that is working fine we can just re-do the same to the other Things.

chilobo replied:
I produced a version which provides seperate things for the main parts of the heatpump- that took some time.
I found a user (Agio Severin Corsten Foundation member) who tested it with two heating circuits, so I would like to commit the version to github. 

I corrected the problems in the README and the xml-files (at leasts I tried to).

Is there a way to erase all the old versions and submit only the new version to github?

You suggested:

> Provide a full usage example based on textual configuration files.

Where do I find these files? Do I just copy what I find in the UI with "Code"? That doesn't provide the format other READMEs provide.

As I understand it the configuration is stored in a Json-Database. Is there a way to extract the requested textual configuration from them?

chilobo replied:
I found it! Instead of using the UI one can place files in $OPENHAB_CONF/things and $OPENHAB_CONF/items
Openhab creates these and shows them with a lock in the UI.

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+
+| Thing              | ThingTypeID | Description                                         |
+| ------------------ | ----------- | --------------------------------------------------- |
+| Lambda Heat Pump   | lambdahp    | A lambda heat pump visible in the local network     |
```
The openHAB model (things, channels etc) is the most important as changing that is breaking and affects users. 
The implementation is less important as it can be changed without breaking (in most cases).
Without looking at the code it would da like you want the shared code to be extracted to a base handler and have the different things extend it.

lsiepel replied:
It's not a question if the bundle is working or not. It is about having proper model that is future fit, we like to prevent breaking changes to users.
I can understand that the the stiebeleletron was used as example as it is somehow similar (modbus) and therefore usefull. But it is also a very old binding (way older then 3 years) that originates from openHAB 2.x and maybe even before that. A lot has changed since.

To me there is no reason to rush, i'd rather have a good modelled binding instead of rushing this and have breaking changes later on. I'm happy to assist if you need any help to change this. 

_PS: Please don't mark something as resolved when it is not yet._

chilobo commented:
I am lost. 
@Java17: The details button does not show what the problem is.
The problem with the DCO Action cannot be resolved by me because I am not a codeowner of openhab?

lsiepel commented:
> I am lost. @Java17: The details button does not show what the problem is. The problem with the DCO Action cannot be resolved by me because I am not a codeowner of openhab?

If you click [details ](https://github.com/openhab/openhab-addons/pull/17595/checks?check_run_id=31989977350) next to the DCO issue, there are 4 commits that lack the signoff. That page also lists some actions to take to fix it. (you have to rewrite the commit message).

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+Other configurations of the E-Manager are not supported (yet).
+
+Up to now only the following configuration is supported:
+- only one heatpump(heatpump1)
```
Add whiteline surrounding lists and add space at the list entry.
```suggestion

- only one heatpump (heatpump1)
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+- only one buffer for the heating (buffer1)
+- only one buffer for the water heating (boiler1)
+
+
```
Remove excess whiteline
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+This extension adds support for the Lambda Heat Pump modbus protocol as provided by
+https://lambda-wp.at/wp-content/uploads/2022/01/Modbus-Protokoll-und-Beschreibung.pdf 
+
+A Lambda Heat Pump has to be reachable within your network. 
```
```suggestion
A Lambda Heat Pump has to be reachable within your network.
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+https://lambda-wp.at/wp-content/uploads/2022/01/Modbus-Protokoll-und-Beschreibung.pdf 
+
+A Lambda Heat Pump has to be reachable within your network. 
+If you plan to use the E-Manager part to hand over your PV excess to the heat pump ask Lambda support to 
+configure it to  
```
```suggestion
If you plan to use the E-Manager part to hand over your PV excess to the heat pump ask Lambda support to configure it to:
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+# Lambda Heat Pump
+
+This extension adds support for the Lambda Heat Pump modbus protocol as provided by
+https://lambda-wp.at/wp-content/uploads/2022/01/Modbus-Protokoll-und-Beschreibung.pdf 
```
```suggestion
<https://lambda-wp.at/wp-content/uploads/2022/01/Modbus-Protokoll-und-Beschreibung.pdf>
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+
+This group contains general operational information about the heat pump.
+
+
```
Remove excess newline
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+| actual-power-consumption         | Number:Power       | true      | [W]      | Power consumption of heatpump 1 (only valid when Betriebsart: Automatik, 0 W otherwise) |
+| power-consumption-setpoint       | Number:Power       | false     | [W]      | Power consumption setpoint for heat pump 1                                              |
+
+
```
Remove excess whiteline
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+| heatpump1-cop                  | Number                    | true      | [0.01%]      | Coefficient of performance                                              | 
+| heatpump1-set-error-quit       | Number                    | false     | [Nr]         | Set Error Quit (1 = Quit all active heat pump errors                    |
+
+
```
Remove excess whiteline
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+| buffer1-actual-low-temperature          | Number:Temperature | true      | [0.1 °C] | Actual temperature buffer low sensor                                   |
+| buffer1-maximum-buffer-temperature      | Number:Temperature | false     | [0.1 °C] | Setting for maximum buffer temperature (min = 25.0°C; max = 65.0°C)    |
+
+
```
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+| heatingcircuit1-room-heating-temperature       | Number:Temperature | rue       | [0.01°C]     | Setting for heating mode room setpoint temperature(min = 15.0°C; max = 40.0 °C) |
+| heatingcircuit1-room-cooling-temperature       | Number:Temperature | true      | [0.01 l/min] | Setting for cooling mode room setpoint temperature(min = 15.0°C; max = 40.0 °C) |
+
+
+
```
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+// Mode of E-Manager has to be switched to AUTOMATIK in the Lambda Heat Pump App
+var int P_Available =  ((Lambda_EMgr_Power_Consumption_Value_as_Number.state as Number) - (PW_Battery.state as Number) - (PW_Grid.state as Number)).intValue 
+    lambdahp_actual_power.sendCommand(P_Available)
+
+
```
```suggestion
```

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/README.md:
```diff
@@ -0,0 +1,174 @@
+This group contains general operational information about the heat pump.
+
+
+| Channel ID                       | Item Type          | Read only | Unit     | Description                                                              |
```
The Unit column can be omitted as the Item Type already describes the type expected. The Unit however is based (or should be) based on the user configured measurement system. (C vs F)

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-types.xml:
```diff
@@ -0,0 +1,27 @@
+		<description>Lambda Heat Pump connected through modbus</description>
+
+		<channel-groups>
+			<channel-group id="generalAmbient" typeId="general-ambient"/>
+			<channel-group id="generalEManager" typeId="general-emanager"/>
+			<channel-group id="Heatpump1" typeId="heatpump1"/>
+			<channel-group id="Boiler1" typeId="boiler1"/>
+			<channel-group id="Buffer1" typeId="buffer1"/>
+			<channel-group id="HeatingCircuit1" typeId="heatingcircuit1"/>
```
Please adjust to the naming convention `lower-case-hypen`
Both the `id` as `typeId`

For reference: https://www.openhab.org/docs/developer/guidelines.html#naming-convention

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+		<label>Ambient Error Number</label>
+		<state readOnly="true">
+			<options>
+				<option value="0">No Error</option>
```
I guess there are more options that can be added? Also usefull for translation.

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+		<label>E-Manager Error Number</label>
+		<state readOnly="true">
+			<options>
+				<option value="0">No Error</option>
```
More options available?

chilobo replied:
No more options, the manual does not state more.

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+	</channel-type>
+
+	<channel-type id="actual-power-type">
+		<item-type unitHint="%">Number:Power</item-type>
```
% does not match a unit the `Number:Power` type. By default it is W. When looking at the handler it also forces `W` so i guess this unithint should be removed.

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+	</channel-type>
+
+	<channel-type id="actual-power-consumption-type">
+		<item-type unitHint="%">Number:Power</item-type>
```
See previous comment. Either remove the unithint (as it does not align with the item-type) or change the item type to Number:Dimenionless (and make it a percentage channel)

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+	</channel-type>
+
+	<channel-type id="power-consumption-setpoint-type">
+		<item-type unitHint="%">Number:Power</item-type>
```
See previous comment. Either remove the unithint (as it does not align with the item-type) or change the item type to Number:Dimenionless (and make it a percentage channel)

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+		<label>Heatpump1 Error Number</label>
+		<state readOnly="true">
+			<options>
+				<option value="0">No Error</option>
```
More options available?

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+		</state>
+	</channel-type>
+
+	<channel-type id="heatpump1-operating-state-type">
```
Applies to all channel types with the id prefix: `heatpump1` the 1 can be removed. As it is the type and the type is used for _all_ heatpumps.

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+
+	<channel-type id="heatpump1-operating-state-type">
+		<item-type>Number</item-type>
+		<label>Heatpump1 Operating State</label>
```
Related to the aboven `heatpump1` comment. Only the label differs, that can be set when defining the channel, so here the label can adjusted to not have a number in it.

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+	</channel-type>
+
+	<channel-type id="heatpump1-t-flow-type">
+		<item-type unitHint="%">Number:Temperature</item-type>
```
Remove unit hint

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+	</channel-type>
+
+	<channel-type id="heatpump1-t-return-type">
+		<item-type unitHint="%">Number:Temperature</item-type>
```
Remove unit hint

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+	</channel-type>
+
+	<channel-type id="heatpump1-vol-sink-type">
+		<item-type unitHint="%">Number:VolumetricFlowRate</item-type>
```
Remove unit hint

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/resources/OH-INF/thing/lambda-heatpump-channel-types.xml:
```diff
@@ -0,0 +1,430 @@
+	</channel-type>
+
+	<channel-type id="heatpump1-t-eqin-type">
+		<item-type unitHint="%">Number:Temperature</item-type>
```
Remove unit hint (before i repeat this again, please adjust all :-) )

lsiepel commented on bundles/org.openhab.binding.modbus.lambda/src/main/java/org/openhab/binding/modbus/lambda/internal/LambdaBindingConstants.java:
```diff
@@ -0,0 +1,104 @@
+    public static final ThingTypeUID THING_TYPE_LAMBDAHP = new ThingTypeUID(BINDING_ID, "lambda-heatpump");
+
+    // Channel group ids
+    public static final String GROUP_GENERAL_AMBIENT = "generalAmbient";
+    public static final String GROUP_GENERAL_EMANAGER = "generalEManager";
+    public static final String GROUP_HEATPUMP1 = "Heatpump1";
+    public static final String GROUP_HEATPUMP1SET = "Heatpump1Set";
+    public static final String GROUP_BOILER1 = "Boiler1";
+    public static final String GROUP_BOILER1MT = "Boiler1Mt";
+    public static final String GROUP_BUFFER1 = "Buffer1";
+    public static final String GROUP_BUFFER1MT = "Buffer1Mt";
+    public static final String GROUP_HEATINGCIRCUIT1 = "HeatingCircuit1";
+    public static final String GROUP_HEATINGCIRCUIT1SETTING = "HeatingCircuit1Setting";
```
Please adjust to the naming convention `lower-case-hypen`

For reference: https://www.openhab.org/docs/developer/guidelines.html#naming-convention

lsiepel reviewed (changes requested):
Thank you for fixing most comments quicly. When looking at the other files i got some new ones. Only the handler is left to review. But that is only usefull after we finished the modelling of things and channels.

chilobo commented:
I think I am finished with the necessary changes.
@ISiepel: Is there anything else missing?

lsiepel commented:
Some commits got accidently added to this PR. Please rebase / undo excess commits

chilobo commented:
@lsiepel 
I did a rebase removed some tracelogs and would ask you to have a look at the project again.

lsiepel commented:
> @lsiepel I did a rebase removed some tracelogs and would ask you to have a look at the project again.

Could you fix the excess commits and conflicts? I guess something went wrong with rebasing.

chilobo commented:
> > @lsiepel I did a rebase removed some tracelogs and would ask you to have a look at the project again.
> 
> Could you fix the excess commits and conflicts? I guess something went wrong with rebasing.

I tried again, but I guess I failed again. In vscode the merging now looks like:
![image](https://github.com/user-attachments/assets/358e1522-6c8a-442e-8f3e-99669cdd0a7c)
... and I have not yet checked if my code is up to date, but I have working copies.
Does this look OK to you?

lsiepel commented:
You have to merge openHAB main into your GitHub main. And from there in your feature branch. It looks like you updated your feature branch straight from openHAB main

chilobo commented:
> You have to merge openHAB main into your GitHub main. And from there in your feature branch. It looks like you updated your feature branch straight from openHAB main

I am lost...
Should I try to do something like:
[https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358](url)

chilobo commented:
> > > You have to merge openHAB main into your GitHub main. And from there in your feature branch. It looks like you updated your feature branch straight from openHAB main
> > 
> > 
> > I am lost... Should I try to do something like: [https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358](url)
> 
> As from there you pull in the changes from upstream straight into your local branch this can be expected.
> 
>     1. You could merge openhab main into your github fork main See here: https://github.com/chilobo/openhab-addons/tree/main is 400+ commits behind.
> 
>     2. Backup the binding folder. Create a new fork / branch etc and in the end a new PR if you can't fix it.
I went back to a really old version, did a git pull and used my local backup to change the project. 
The old pull request was closed - that is OK with me.
Should I delete this fork now?

Then I pushed my local fork to github and opened up a new pull request:
 [ [modbus.lambda] Lambda Heat Pump #17846](https://github.com/openhab/openhab-addons/pull/17846)

But ...
I changed CODEOWNERS, but forgot to change pom.xml in the bundles directory, so the compilation failed. 
Now I added 
  <module>org.openhab.binding.modbus.lambda</module>
in the bundles\pom.xml and pushed it to github.
Will the compilation start automagically or do I have to do something else?

chilobo commented:
I fixed the DCO failures and the compilations worked out. 
So: When you find time, please look at the new pull request.
Should I delete the old fork?

lsiepel commented:
I'll look into the new PR, let's continue there.

chilobo commented:
The new PR ist waiting for your approval or comments since about 6 weeks.....
